danielsimig/metaseq
# Compute/use data pruning metrics in metaseq

**Patch Description**
Adding changes to log_model_outputs to compute data pruning metrics and use computed metrics to filter data while pruning

**Testing steps**
To compute metrics (metrics that are dependent on running the model pipeline), you can run something like:
```
 metaseq-train --task streaming_language_modeling /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321 --save-dir /checkpoint/ktirumala/tempdebug/ --arch transformer_lm --share-decoder-input-output-embed --dropout 0.1 --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0 --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 --tokens-per-sample 512 --sample-break-mode none --tokens-per-sample 2048 --update-freq 1 --fp16 --log-format json --log-interval 1 --validate-interval-updates 1000 --save-interval-updates 100 --keep-interval-updates 1 --vocab-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-vocab.json --merges-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-merges.txt --valid-subset valid/redditflattened --batch-size 2 --compute-data-pruning-metrics ppl --compute-data-pruning-metrics-savedir /checkpoint/ktirumala/lm_data_pruning_v2/
```

Right now, the supported metrics for `--compute-data-pruning-metrics` are `ppl` and `ssl_prototypes` 

To use data pruning metrics, you should create a `.jsonl` file where each line is:
```
{
            "name": dataset_name, 
            "index": index in dataset jsonl file, 
            "metric": metric value
}
```

Then you can run 
```
metaseq-train --task streaming_language_modeling /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321 --save-dir /checkpoint/ktirumala/tempdebug/ --arch transformer_lm --share-decoder-input-output-embed --dropout 0.1 --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0 --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 --tokens-per-sample 512 --sample-break-mode none --tokens-per-sample 2048 --update-freq 1 --fp16 --log-format json --log-interval 1 --validate-interval-updates 1000 --save-interval-updates 100 --keep-interval-updates 1 --vocab-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-vocab.json --merges-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-merges.txt --valid-subset valid/redditflattened --batch-size 2 --use-data-pruning-metrics --use-data-pruning-metrics-filepath /checkpoint/ktirumala/tempdebug/metrics.jsonl
```
